File: app/src/main/java/com/dynamo/spacex/shared/TransientDataProvider.java
package com.dynamo.spacex.shared;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class TransientDataProvider {

    private ConcurrentMap<Class, UseCase> transientData = new ConcurrentHashMap<>();

    @Inject
    public TransientDataProvider() { }

    public void saveUseCase(UseCase useCase) {
        final Class<? extends UseCase> useCaseClass = useCase.getClass();
        transientData.put(useCaseClass, useCase);
    }

    public <T extends UseCase> T getUseCase(Class<T> useCaseClass) {
        final UseCase removedUseCase = transientData.remove(useCaseClass);
        return useCaseClass.cast(removedUseCase);
    }
}

File: app/src/main/java/com/dynamo/spacex/shared/TransientDataProvider.kt
package com.dynamo.spacex.shared

import java.util.concurrent.ConcurrentHashMap

class TransientDataProvider {
    private val transientData = ConcurrentHashMap<Class<out UseCase>, UseCase>()

    fun saveUseCase(useCase: UseCase) {
        val useCaseClass: Class<out UseCase> = useCase.javaClass
        transientData[useCaseClass] = useCase
    }

    fun getUseCase(useCaseClass: Class<out UseCase>): UseCase? {
        return transientData.remove(useCaseClass)
    }

    fun containsUseCase(useCaseClass: Class<out UseCase>): Boolean {
        return transientData.containsKey(useCaseClass)
    }
}